File: readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // waits for input
    string text = get_string("Text: ");

    // starts the parsing of the text
    int num_letters = count_letters(text);
    int num_words = count_words(text);
    int num_sentences = count_sentences(text);

    // aplies the formula
    float L = (float) num_letters / (float) num_words * 100;
    float S = (float) num_sentences / (float) num_words * 100;
    int index = round(0.0588 * L - 0.296 * S - 15.8);

    // decides the output based on the result of the formula
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
    return 0;
}

int count_letters(string text)
{
    int length = strlen(text);
    int i = 0, letters = 0;

    // starts counting the number of letters/numbers
    while (i < length)
    {
        if (isalnum(text[i]))
        {
            letters++;
        }
        i++;
    }
    return letters;
}

int count_words(string text)
{
    int length = strlen(text);
    int i = 0, words = 0;

    // starts counting the number of words (separated by spaces)
    while (i < length)
    {
        if (isspace(text[i]))
        {
            words++;
        }
        i++;
    }
    return words + 1;
}

int count_sentences(string text)
{
    int length = strlen(text);
    int i = 0, sentences = 0;

    // starts counting the number of sentences (separated by '.', '!' or '?')
    while (i < length)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
        i++;
    }
    return sentences;
}

File: sentimental-readability/readability.py
import re


def main():
    # Asks for input
    text = input("Text: ")

    # Starts parsing the text
    num_letters = count_letters(text)
    num_words = count_words(text)
    num_sentences = count_sentences(text)

    # Aplies the formula
    L = num_letters / num_words * 100
    S = num_sentences / num_words * 100
    index = 0.0588 * L - 0.296 * S - 15.8

    # Decides the output
    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {index:.0f}")


# Counts the number of letters in the text
def count_letters(text):
    letters = re.findall("\w", text)
    letters = list(filter(("_").__ne__, letters))
    return len(letters)


# Counts the number of words in the text
def count_words(text):
    words = re.split("\s", text)
    return len(words)


# Counts the number of sentences in the text
def count_sentences(text):
    sentences = re.split("[.!?]", text)
    return len(sentences) - 1


if __name__ == "__main__":
    main()